### ISSUE TITLE
Add thread-safe spawn extension with redirection support

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/libcx#49]
As the only way to make a child process use specific file handles for its standard I/O is to redirect these handles in the parent before launching the child, a race condition between different threads of the parent process that might want to do it at the same time arises. A real life use case is here: http://trac.netlabs.org/rpm/ticket/280.

This function (standard I/O redirection) is very widely used in many applications. Hence is the idea to provide a `spawn` extension in LIBCx that will provide such a redirection in a thread-safe manner. This extension could be the used in all those applications instead of providing quite lengthy custom logic in each of them.

**Comment by dmik** [comment]
This is the proposed interface:
```c
/**
 * Starts a child process using the executable specified in @a name.
 *
 * Program arguments are specified in @a argv whose last element must be NULL.
 * The filst element of this array should be the program name, by convention,
 * and must always be non-NULL. The current working directory of the new
 * process is specified with @a cwd. NULL will casuse the new process to
 * inherit the working directory of the starting process. The environment of
 * the new process is either given in the @a envp array whose last element must
 * be NULL, or inherited from the starting process if @a envp is NULL.

 * If the starting process wants to redirect standard I/O streams of the child
 * process, it should specify the new handles in the @a stdfds array (which
 * otherwise should be NULL to indicate that the child process should inherit
 * standard I/O handles of its parent). This array, if not NULL, must contain 3
 * elements that represent LIBC file descriptors to be used by the child's
 * stdin, stdout and stderr streams, respectively. Some integer values have a
 * special meaning when used in this array and therefore these values are never
 * interpreted as file descriptors:
 *
 * - 0 indicates that the respective stream should not be redirected but
 * instead inherited from the parent.
 *
 * - 1 indicates that the stderr stream should be redirected to the stdout
 * stream. A value of 1 can only be put in the last element of @a stdfds (with
 * an index of 2) and is invalid otherwise.
 *
 * - 2 is reserved for future use and is currently interpreted as an invalid
 * value.
 *
 * In addition to the standard `P_*` constanst for the @a mode argument, the
 * following new values are recognized:
 *
 * - P_2_NOINHERIT causes `spawn2` to prevent inheritance of all file handles
 * of the starting process by the child process.
 *
 * Note that to guarantee full thread safety, only one thread may execute
 * `spawn2` at a time, all other threads of the calling process will be blocked
 * for the duration of the call. For this reason using the P_WAIT flag is not
 * allowed with `spawn2` and passing it will immediately result in an error
 * with `errno` set to E_INVAL.
 *
 * The function returns the process ID of the child process (P_NOWAIT, P_DEBUG,
 * P_SESSION and P_PM), or zero (P_UNRELATED) if successful. On error it
 * returns -1 and sets `errno` to the respective error code.
 *
 * Except for what is described above, the `spawn2` function behaves
 * essentially the same as the standard `spawnvpe` LIBC function. Please
 * consult the `spawnvpe` manual for more information
 */
int spawn2(int mode, const char *name, const char * const argv[],
           const char *cwd, const char * const envp[], int stdfds[3]);
```

Among other things, it supports setting the current directory for the new process. I wonder if we should return ENOENT if `cwd` is given but changing to that directory fails. This seems logical.

**Comment by dmik** [comment]
Provided the initial implementation of the above interface. Thread safety isn't there yet as I need a good test case for it. Also some more test cases are needed to cover the full functionality (e.g. P_2_NOINHERIT and also stdin redirection; note that stdout/stderr redirection testing is already there).

**Comment by dmik** [comment]
I wrote a test that starts 25 threads each calling `spawn2` with I/O redirection. This test expectedly fails with no synchronization. When using `DosEnterCritSec` in `spawn2` to protect code that duplicates file handles and changes the current dir, it starts working however gets eventually stuck if the number of concurrent threads is greater than 5. I really don't want to debug it: there is a lot of LIBC function calls from within `DosEnterCritSec` and some of them use mutexes for synchronization. More over, `spawnvpe` itself eventually uses a mutex to protect some internal LIBC process management structures. So a situation is likely when some thread being inside a direct call to `spawnvpe` gets blocked by another thread calling `DosEnterCritSect` by `spawn2` while holding that mutex. This will result in a dead lock when `spanwn2` calls `spanwnvpe` because it will require the other thread to release this lock and that won't happen because that thread is blocked.

So we have to give up the idea of using `DosEnterCritSec` for that. Instead, we will have to use a regular mutex that will serialize just `spawn2` calls themselves. I already tried it and that works really well for any number of threads I tried. There is another problem though: when the number of threads is about 50 or so `spawnvpe` starts failing with errno set to zero. However, this looks like an irrelevant kLIBC bug to me — most likely, `DosExecPgm` starts failing (50 child processes in less than a second is quite a bit) or some other kLIBC misbehavior. I will leave it for now until we get a real use case that wants that many children (which I doubt will happen soon).

There is an obvious disadvantage of using a mutex instead of `DosEnterCritSec`: any other thread doing some std I/O while `spawn2` is executed will easily screw it up. However, this sort of race is pretty much high level and things should be synchronized on the app level in this case: it's basically the same race as when two threads simply write to std I/O w/o synchronization, they will lend up with misplaced output. So not ideal but I think we can live with that.

Now I will try to use `spawn2` from Python and run JS JIT tests against it — if they won't hang any more (see http://trac.netlabs.org/rpm/ticket/280 and https://github.com/bitwiseworks/mozilla-os2/issues/247#issuecomment-350467681), I will consider this as done.

**Comment by dmik** [comment]
I made python use new `spawn2` but unfortunately something goes wrong, JIT tests screw it up. From what I see it starts tests on thread 3 but still tries to write something to stdout on thread 1. Which means that there is a race not between two threads doing `spawn2` but instead between one writing to stdout and another one doing `spawn2`. A spawn-own mutex is of course not to protect from that. I have to figure out that to do.

**Comment by dmik** [comment]
Well, we are in trouble here. There is no universal solution, unfortunately. Even though using `DosEnterCritSec` works under some circumstances (I got like 2k tests run with it), it will eventually hang. There are too many potential deadlocks: e.g. even kLIBC `__dup2` requests a mutex; so if there is another thread doing `dup2` while the `spawn2`thread enters the critical section, then the latter thread will get stuck forever in its own `dup2` call because the former thread holding `dup2` mutex will never finish. And there is a dozen places like that.

The only proper solution here is to move this dup stdio functionality into the OS/2 kernel (where it could creating these duplicates under a loc preventing other `DosDupHandle` calls from this process) but of course this is hardly possible.

So the only way to go here if we want to achieve thread-safety on a generic level within `spawn2` is to use an intermediate process (via a special helper EXE) that will do whatever it needs for redirection w/o affecting handles in the starting process. This will double the number of processes and in this regard is equivalent to using `fork`, the only essential difference is that copying data segments and other magic is not done which'll still make it faster and more efficient than `fork`.

I guess I will introduce a flag, P_2_THREADSAFE, that will activate this mode. And w/o this flag, `spawn2` will behave w/o any special synchronization, just like the original `spawnvpe` and friends.

**Comment by dmik** [comment]
Luckily, for the Python/JIT test case I found another way to achieve a usable result, see http://trac.netlabs.org/rpm/ticket/280#comment:11. So the P_2_THREADSAFE approach will be postponed till later (it's not a 5 minute work).

**Comment by dmik** [comment]
The above result is not really usable, see https://github.com/bitwiseworks/mozilla-os2/issues/247#issuecomment-352456327. Also, it seems that things like `pipe()` are not thread-safe in kLIBC. At least, when running my test case with the whole `spawn2` body protected by a separate mutex and unprotected pipe creation on worker threads many times, I get EBADF from time to time (coming from `read` on the parent side) or EINVAL (`dup` within `spawn2` itself). I'm not sure if it's kLIBC `pipe()` that is not thread safe, or `DosCreatePipe` itself, but anyway, everything is too fragile for real multithreading over there. The test case becomes rock solid *only* if I properly protect the whole sequence starting from creating a pair of pipes and ending with writing and reading from them (with `spawn2` in between). So again, we have to give up the idea of making `spawn2` thread-safe by itself. An application modification is necessary to make it really thread-safe.

As for Python, as the above link shows, it gets screwed up because of `fprintf`. This particular problem may be solved by locking std streams with `STREAM_LOCK` (this is what all FILE stream APIs internally do). This should solve the problem with the JIT framework but direct use of the file handle is still possible (with e.g. `read` or `write` or `pipe` LIBC calls) so it's not a universal solution either.

**Comment by dmik** [comment]
@komh writes here https://github.com/bitwiseworks/mozilla-os2/issues/247#issuecomment-352720239:

> EPIPE occurs if trying to write to the pipe whose other end was closed. In this case, a thread terminates unless a signal handler to continue was installed.
EBADF occurs if trying to read/write from/to the fd which was closed or not allocated regardless of pipes.
And I agree with your guess about EBADF caused by multi-threads.
For this, how about using global mutex for dup(), dup2(), spawn...(), and exec...() with wrapping them.
And use DosEnterCritSec()/DosExitCritSec() pair within the mutex of spawn2(). With this, it will be possible to avoid dead-lock due to internal mutexes in kLIBC.

Sure, using a global mutex would solve the problem but this is too much work and you need to wrap not only the listed functions but *all* functions taking an `int fd` as an argument (i.e. read, write, ftruncate etc etc — you name it). That's just too much work and it will give a significant performance slow down: imagine a mutex requested each time `read` is called — acquiring a mutex is not cheap at all. Yes, we could save some cycles by using simple mutexes (an atomic read/write + a spin loop) hoping that collisions don't happen often but it's still some loss + a lot of work to create/maintain wrappers. An approach with a wrapper EXE looks much simpler and it will only have a performance penalty in case of `spawn2` usage, not on each `read` or `write`.

**Comment by dmik** [comment]
BTW, for the wrapper approach to work correctly in P_NOWAIT mode, I need to exchange data between the wrapper and the `spawn2` calling process using LIBCx shared memory — in order to feed it the execution result of the real child (waiting on a grand-child with retrieving its return code and status is not possible on OS/2). This means that I will also need to override `waitpid` to free this shared memory area.

Beside the above, the wrapper mode is done and appears to work fine.

**Comment by dmik** [comment]
Another important thing with wrapping is that we need to give the real (grand-child) process ID to the calling process so that things that require the real PID like giving access to the shared memory could work correctly.

**Comment by komh** [comment]
Hmm... I don't think all funcs using fd should be wrapped. Because global mutex is used to avoid dead-lock with in kLIBC. That is, funcs which are used in spawn2() are enough. For other threads, DosEnterCritSec() after global mutex acqusition will prevent them from being confused.

Anyway, if you solved with your wrapper approach, it's ok. Good job!

**Comment by dmik** [comment]
@komh every func that might potentially use stdio handles (0, 1, 2 or their dupes) needs to be serialized WRT `spawn2` (this includes `read/write`). kLIBC uses several mutexes in different places, there is no single mutex for every fd access. And currently there is no any enclosing locks in `read` and friends: it's barely a call to `DosRead` in case of simple file handles (yes, there is a `gmtx` lock to handle internal structure retrieval, but it's released before `DosRead`). And in order to use `DosEnterCritSec` without risking to get a deadlock, we must acquire and hold not only `gmtx` but *all* existing kLIBC mutexes before calling `DosEnterCritSec`. More over, we must acquire them in some well defined order (and all other code should maintain the very same order as well, otherwise it's also a door to a deadlock). In other words, this is barely possible in practice.

**Comment by komh** [comment]
@dmik I thought of DosDupHandle() as well to avoid fd problems. But I've confirmed that __spawnve() also uses mutex related to fds(gmtx) in order to inherit fds. After all, I've concluded it's not possible to avoid potential dead-locks with DosEnterCritSec().
Maybe, rather it may be easier to implement spawn2() with OS/2 APIs such as DosDupHandle(), DosExecPgm() and DosEnterCritSec() without kLIBC functions at all.

**Comment by StevenLevine** [comment]
@komh I don't think this approach can work.  The problem is we need to inherit kLIBC file handles, not OS/2 file handles.  The handle could be a socket handle.

FWIW, it is possible to acquire a known set of mutexes, while avoiding deadlock and without the explicit cooperation of other code.  However, it requires a relatively complex retry loop.  In theory, the retry loop could retry forever. In practice, this does not occur.

Another FWIW,  one needs to keep in mind that DosEnterCritSec is not a substitute for mutexes.  I view DosEnterCritSec as as more of a temporary priority boost that allows the thread to quickly finish the work it needs to do and release the mutexes it holds.

**Comment by dmik** [comment]
This wrapper approach turned to be VERY COMPLEX when it came to edge cases like reporting error conditions or passing back signals from the crashed child to the wrapper caller. I had several rounds of creating various workarounds but none could survive *all* tests. 

You have P_OVERLAY in spawn that forwards signals from the wrapped child to the wrapper's parent process very well. However, P_OVERLAY makes spawn never return to the caller on success and this makes it impossible to report spawn failures like `file not found` etc since the parent cannot both wait for a result and not wait for it. For LIBCx-based processes this is easily solvable in DLL init but this approach doesn't work for other EXEs (like cmd.exe).

A proper way would be to first call spawn with (P_NOWAIT), report the result of spawn to the parent and then do what P_OVERLAY does to finally report the child rc and signals to the parent. But there is no public function in kLIBC for that. The logic is pretty simple and can be grabbed from `emx/src/lib/sys/__spawnve.c` but we miss one __libc export for that: `__libc_spmTerm`, although it is defined in the public header (`<InnoTekLIBC/sharedpm.h>`).

Exporting it from our RPM kLIBC is of course very easy but this requires to change the naming/versioning scheme as it won't be a drop-in replacement any more... We *do need* a kLIBC fork with our own naming scheme for a long-long time, but getting distracted by that task is another week or so (and I'm not sure I can hold that many nested tasks). So far I'm confused.

**Comment by komh** [comment]
Another idea. Just passing fds to be duplicated via LIBCX, and dup()ing at startup in a child process. LIBCX has a hook for main(). Even if a process not linked against LIBCS, it seems that it's possible to hook via SYS_DLLS/LoadPerProcess.

**Comment by dmik** [comment]
@komh yes, that's also an option but I think that SYS_DLL is generally evil. Anyway, I finished and committed the thread-safe code in c37e695870f522f5db6ae0fcf7c8a1ba30608a76. I didn't have to hack kLIBC around or reduce the planned functionality. I just used `raise` to pass the child result back to the parent. It has some limitations (when debugging) but they are not too relevant so normal apps won't be affected. More over, I even have the ID of the grand child so overriding `waitpid` is really piece of cake now (all data is there). Not needed now so postponed in #51.

The wrapper code It's rock solid, it survives a very tight test case loop that starts 25 processes within a few dozen milliseconds on 4 cores in SMP mode (so things get really mixed up there). Now I need to test it with JIT but I'll do it within the relevant ticket (see above) as it's totally off-topic here and the code is worth being in LIBCx on its own anyway. Closing this one.
